erizo/src/erizo/rtp/QualityFilterHandler.cpp

```diff
@@ -0,0 +1,130 @@
+#include "rtp/QualityFilterHandler.h"
+
+#include "./WebRtcConnection.h"
+#include "lib/ClockUtils.h"
+#include "rtp/RtpUtils.h"
+
+namespace erizo {
+
+DEFINE_LOGGER(QualityFilterHandler, "rtp.QualityFilterHandler");
+
+QualityFilterHandler::QualityFilterHandler()
+  : connection_{nullptr}, enabled_{true}, initialized_{false},
+  target_spatial_layer_{0}, target_temporal_layer_{0},
+  video_sink_ssrc_{0}, video_source_ssrc_{0}, last_ssrc_received_{0} {}
+
+void QualityFilterHandler::enable() {
+  enabled_ = true;
+}
+
+void QualityFilterHandler::disable() {
+  enabled_ = false;
+}
+
+void QualityFilterHandler::handleFeedbackPackets(std::shared_ptr<dataPacket> packet) {
+  RtpUtils::forEachRRBlock(packet, [this](RtcpHeader *chead) {
+    RtpUtils::updateREMB(chead, 400000);
+
+    RtpUtils::forEachNack(chead, [this, chead](uint16_t seq_num, uint16_t plb) {
+      SequenceNumber result = translator_.reverse(seq_num);
+      if (result.type == SequenceNumberType::Valid) {
+        chead->setNackPid(result.input);
+      }
+    });
+  });
+}
+
+void QualityFilterHandler::read(Context *ctx, std::shared_ptr<dataPacket> packet) {
+  if (enabled_) {
+    handleFeedbackPackets(packet);  // TODO(javier) remove this line when RTCP termination is enabled
+
+    // TODO(javier): Handle RRs and NACKs and translate Sequence Numbers?
+  }
+
+  ctx->fireRead(packet);
+}
+
+void QualityFilterHandler::checkLayers() {
+  int new_spatial_layer = quality_manager_->getSpatialLayer();
+  if (new_spatial_layer != target_spatial_layer_) {
+    sendPLI();
+    target_spatial_layer_ = new_spatial_layer;
+  }
+  int new_temporal_layer = quality_manager_->getTemporalLayer();
+  target_temporal_layer_ = new_temporal_layer;
+}
+
+void QualityFilterHandler::checkSSRCChange(uint32_t ssrc) {
+  if (last_ssrc_received_ != ssrc) {
+    translator_.reset();
+  }
+  last_ssrc_received_ = ssrc;
+}
+
+void QualityFilterHandler::sendPLI() {
+  getContext()->fireRead(RtpUtils::createPLI(video_sink_ssrc_, video_source_ssrc_));
+}
+
+void QualityFilterHandler::write(Context *ctx, std::shared_ptr<dataPacket> packet) {
+  RtcpHeader *chead = reinterpret_cast<RtcpHeader*>(packet->data);
+  if (!chead->isRtcp() && enabled_ && packet->type == VIDEO_PACKET) {
+    RtpHeader *rtp_header = reinterpret_cast<RtpHeader*>(packet->data);
+
+    checkLayers();
+
+    uint32_t ssrc = rtp_header->getSSRC();
+    uint16_t sequence_number = rtp_header->getSeqNumber();
+
+    if (!packet->belongsToSpatialLayer(target_spatial_layer_)) {
+      if (ssrc == video_sink_ssrc_) {
+        translator_.get(sequence_number, true);
+      }
+      return;
+    }
+
+    checkSSRCChange(ssrc);
+    rtp_header->setSSRC(video_sink_ssrc_);
+
+    if (packet->compatible_spatial_layers.back() == target_spatial_layer_ && packet->ending_of_layer_frame) {
```

Comment:
If I understand this right, if we put this after current line 95, we won't be setting the marker for packets that we're going to discard anyway